driver/ppm_fillers.c

```diff
@@ -723,106 +723,116 @@ if (append_cgroup(#_x, _x ## _subsys_id, args->str_storage + STR_STORAGE_SIZE -
 #endif
 
 /* Takes in a NULL-terminated array of pointers to strings in userspace, and
- * concatenates them to a single \0-separated string. Return the length of this
- * string, or <0 on error */
-int accumulate_argv_or_env(const char __user * __user *argv,
-				  char *str_storage,
-				  int available)
+ * concatenates them to a single \0-separated string. Return the length of these
+ * strings with the final '\0' included.
+ */
+int accumulate_argv_or_env(const void __user * argv,
+				  char *str_storage)
 {
 	int len = 0;
-	int n_bytes_copied;
-
-	if (argv == NULL)
-		return len;
+	int ret = 0;
+	const char __user * p = NULL;
 
 	for (;;) {
-		const char __user *p;
+
+		if (argv == NULL)
+			break;
 
 		if (unlikely(ppm_get_user(p, argv)))
-			return PPM_FAILURE_INVALID_USER_MEMORY;
+		{
+			/* We return what we read until now */
+			break;
+		}
 
 		if (p == NULL)
 			break;
 
-		/* need at least enough space for a \0 */
-		if (available < 1)
-			return PPM_FAILURE_BUFFER_FULL;
-
-		n_bytes_copied = ppm_strncpy_from_user(&str_storage[len], p,
-						       available);
-
-		/* ppm_strncpy_from_user includes the trailing \0 in its return
-		 * count. I want to pretend it was strncpy_from_user() so I
-		 * subtract off the 1 */
-		n_bytes_copied--;
-
-		if (n_bytes_copied < 0)
-			return PPM_FAILURE_INVALID_USER_MEMORY;
-
-		if (n_bytes_copied >= available)
-			return PPM_FAILURE_BUFFER_FULL;
+		/* ppm_strncpy_from_user includes the trailing \0 */
+		ret = ppm_strncpy_from_user(&str_storage[len], p,
+						       STR_STORAGE_SIZE-len);
+		if(ret < 0)
+		{
+			/* We ignore the failed read. We will try to read from the same position in
+			 * the next iteration.
+			 */
+			ret = 0;
+		}
 
-		/* update buffer. I want to keep the trailing \0, so I +1 */
-		available   -= n_bytes_copied+1;
-		len         += n_bytes_copied+1;
+		len += ret;
+		if(len >= STR_STORAGE_SIZE)
+		{
+			len = STR_STORAGE_SIZE;
+			break;
+		}
 
-		argv++;
+		argv += sizeof(argv);
 	}
 
+	if(len>0)
+	{
+		str_storage[len-1] = '\0';
+	}
+	else
+	{
+		str_storage[0] = '\0';
+	}
```

Comment:
It looks like `len` is still sent to 0 in this case while there is one character, while in the case above `len` includes the terminator. It may be fine but could you please double check? 🙏

Reply:
that's a good point, we do that because after this call we call a `strnlen` on `args->str_storage` and we want the strnlen to return `0`. This is the code flow:

```c
				args_len = accumulate_argv_or_env((const char __user *)val,
							   args->str_storage);
		}

		if (args_len == 0)
			*args->str_storage = 0;

		exe_len = strnlen(args->str_storage, args_len);
		// we add the `\0` terminator
		if (exe_len < args_len)
			++exe_len;
```

as you can see also before this patch if `if (args_len == 0)` we `*args->str_storage = 0;` without increasing the len because we just want to clear the buffer before the `strnlen`

Reply:
so we shouldn't need the `len++` inside `accumulate_argv_or_env`

Reply:
Thank you for checking!